Auto-dismiss notifications when background scan disabled

teal-bauer opened the pull request:
## Summary
This PR improves notification management by automatically dismissing the persistent notification when users disable background scanning, eliminating the need for manual dismissal. Additionally, notifications are now only shown when there are saved scooters to connect to.

## Changes
- **Automatic notification dismissal**: When background scanning is toggled off, the service switches to background mode and dismisses the notification programmatically
- **Smart notification management**: Notifications only appear when saved scooters are configured for auto-connect
- **Dynamic foreground mode**: Service automatically switches between foreground and background modes based on scanning state and scooter availability
- **Scooter addition handling**: When a scooter is added while background scanning is enabled, the notification is automatically shown

## Technical Details
Previously, Android foreground services required a persistent notification that users had to dismiss manually. This PR addresses that by:
1. Using `setAsBackgroundService()` when scanning is disabled or no scooters exist
2. Using `setAsForegroundService()` only when scanning is enabled and scooters are configured
3. Dismissing notifications programmatically after switching to background mode

## Testing
Tested on Android (Pixel 8, API 36):
- ✅ No notification when background scan is off
- ✅ No notification when background scan is on but no scooters configured
- ✅ Notification appears when scooter is added
- ✅ Notification dismisses when toggling background scan off
- ✅ Notification shows scooter status when connected

teal-bauer committed:
feat: auto-dismiss notifications when background scan disabled

Automatically dismiss the persistent notification when users disable
background scanning instead of prompting them to dismiss it manually.
The service now switches between foreground and background modes based
on scanning state, allowing proper notification dismissal on Android.

Additionally, notifications are now only shown when there are saved
scooters configured for auto-connect. This prevents unnecessary
notifications when the app has nothing to connect to.

Key changes:
- Switch service to background mode when scanning disabled
- Dismiss notification programmatically after mode switch
- Only show notification when saved scooters exist
- Auto-enable notification when scooters are added
- Update initial notification text to reflect transient state

teal-bauer commented:
> You maniac actually did it...cannot test until tomorrow, but does this affect the home screen/lock screen widgets on any platform? They also rely on the background service being available.

Oh, I didn't realize they need the background service, so I haven't tested that - but as soon as there is a saved scooter and the background connection setting is enabled, the notification persists so it _should_ not change anything.

teal-bauer committed:
refactor: reduce code duplication in background service

TheFreal reviewed (changes requested):
Tested for a while today, seems like this change causes major reliability issues with the home screen widget. Could you please have another look?

teal-bauer commented:
Yeah, I'm on it, just wasn't at my proper dev computer for a few days. I'll ping you when a new iteration is ready to test, I think checking for widget instantiation ought to do it maybe...

teal-bauer committed:
feat: keep service alive for widgets

- Service stays in foreground mode always (ensures widgets work reliably)
- Show notification when background scanning is ON OR widgets are installed
- Dismiss notification only when background scanning is OFF AND no widgets installed
- Remove service mode switching logic (_updateServiceMode)

teal-bauer commented:
With these changes the service doesn't get changed, only the notification gets removed if background service is toggled off AND no widgets are installed.

TheFreal committed:
Merge branch 'main' into feat/programmatic-notification-dismissal

TheFreal commented:
I've removed all home screen widgets and disabled background connections, and am getting this notification on app launch:
<img width="1080" height="845" alt="Screenshot_20251031-125432" src="https://github.com/user-attachments/assets/384ccb1c-4898-4d4f-b6be-95ff9297ad31" />

teal-bauer commented:
Superseded by #141, which splits the foreground-service notification onto a silent `unu_service` channel and adds proper promote/demote lifecycle with a 15-minute timeout. That addresses both the dismissal goal here and the widget-action regression TheFreal hit. Closing.
